test/jasmine/tests/calcdata_test.js

```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
+                    categorymode: 'category ascending',
+                    categorylist: ['b','a','d','e','c'] // These must be ignored. Alternative: error?
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(11);
+                expect(gd.calcdata[0][1].y).toEqual(13);
+                expect(gd.calcdata[0][2].y).toEqual(15);
+                expect(gd.calcdata[0][3].y).toEqual(14);
+                expect(gd.calcdata[0][4].y).toEqual(12);
+            });
+
+            it('should output categories in ascending domain alphanumerical order, excluding undefined', function() {
+
+                Plotly.plot(gd, [{x: ['c',undefined,'e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'category ascending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(11);
+                expect(gd.calcdata[0][1].y).toEqual(15);
+                expect(gd.calcdata[0][2].y).toEqual(14);
+                expect(gd.calcdata[0][3].y).toEqual(12);
+            });
+        });
+
+        describe('codomain numerical category ordering', function() {
+
+            it('should output categories in ascending codomain numerical order', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'value ascending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(11);
+                expect(gd.calcdata[0][1].y).toEqual(12);
+                expect(gd.calcdata[0][2].y).toEqual(13);
+                expect(gd.calcdata[0][3].y).toEqual(14);
+                expect(gd.calcdata[0][4].y).toEqual(15);
+            });
+
+            it('should output categories in descending codomain numerical order', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'value descending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(15);
+                expect(gd.calcdata[0][1].y).toEqual(14);
+                expect(gd.calcdata[0][2].y).toEqual(13);
+                expect(gd.calcdata[0][3].y).toEqual(12);
+                expect(gd.calcdata[0][4].y).toEqual(11);
+            });
+
+            it('should output categories in descending codomain numerical order, excluding nulls', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,null,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'value descending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(15);
+                expect(gd.calcdata[0][1].y).toEqual(14);
+                expect(gd.calcdata[0][2].y).toEqual(12);
+                expect(gd.calcdata[0][3].y).toEqual(11);
```

Comment:
Clarifications:

- When either the category or the value of a trace point is null, it gets mapped to {x: false, y: false} by the current production code, rather than a full disappearance, I haven't changed it.
- In case a category is enlisted in categoryarray but has no point for it in the trace: the x values in the {x,y} index tuples will correctly 'jump over' the category which doesn't have a data point. However there's no artificially generated point for such a category so here as well, we don't exactly have a null value (simply, there won't be a tuple with x as a missing category).

As I'm looking at the generated plots as well, I found something unexpected: when the first N or last N categories in 'categoryarray' have no corresponding point(s) in the trace, they're chopped from the axis. My guess is that it is to do with the automatic zooming onto the actual range of the data, usually done in the case of continuous axes. 

I believe it's not the desired behavior for categorical axes. If it's indeed a range focus thing, then it's a preexisting condition that arose now because in the past, the axis ticks were driven by the trace points, so trivially, there used to be at least one point for each category. Should I work on solving it, or should we optimize for as quick merging of the baseline functionality as possible?

Here's the data:
```
                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
                    type: 'category',
                    categorymode: 'array',
                    categorylist: ['s','y','b','x','a','d','z','e','c', 'q', 'k']
                }});
```
and here's the chart: 
![image](https://cloud.githubusercontent.com/assets/1548516/14355491/63c84578-fce1-11e5-8bf5-bebf04b0affa.png)